# Atividade 1 — Exercícios com os comandos básicos de pandas para EDA


Responda as perguntas abaixo no notebook (cada resposta deve ter ao menos uma célula de código que produza o output e, logo abaixo, uma célula de texto com a sua interpretação).

## 1) Carregue o dataset e mostre as primeiras 10 linhas.

In [6]:
import pandas as pd
df = pd.read_parquet('emendas_tratamento.parquet')
df.head(10)

,Código da Emenda,Ano da Emenda,Tipo de Emenda,Código do Autor da Emenda,Nome do Autor da Emenda,Número da emenda,Localidade de aplicação do recurso,Código Município IBGE,Município,Código UF IBGE,...,Nome Ação,Código Plano Orçamentário,Nome Plano Orçamentário,Valor Empenhado,Valor Liquidado,Valor Pago,Valor Restos A Pagar Inscritos,Valor Restos A Pagar Cancelados,Valor Restos A Pagar Pagos,Código da Emenda_1
0,Sem informação,2014,Emenda Individual - Transferências com Finalid...,NaN,None,S/I,ALVORADA - RS,4300604,ALVORADA,4300000,...,APOIO A POLITICA NACIONAL DE DESENVOLVIMENTO U...,0,APOIO A POLITICA NACIONAL DE DESENVOLVIMENTO U...,700000.00,0.0,0.0,0.0,690900.0,9100.00,Sem informação
1,Sem informação,2014,Emenda Individual - Transferências com Finalid...,NaN,None,S/I,SANTA ISABEL DO PARÁ - PA,1506500,SANTA ISABEL DO PARÁ,1500000,...,APOIO A PROJETOS DE INFRAESTRUTURA TURISTICA,0,APOIO A PROJETOS DE INFRAESTRUTURA TURISTICA -...,487500.00,0.0,0.0,0.0,487500.0,0.00,Sem informação
2,Sem informação,2014,Emenda Individual - Transferências com Finalid...,NaN,None,S/I,BARRA BONITA - SC,4202099,BARRA BONITA,4200000,...,ESTRUTURACAO DA REDE DE SERVICOS DE ATENCAO PR...,0,ESTRUTURACAO DA REDE DE SERVICOS DE ATENCAO BA...,100000.00,0.0,0.0,0.0,0.0,100000.00,Sem informação
3,Sem informação,2014,Emenda Individual - Transferências com Finalid...,NaN,None,S/I,ARARAQUARA - SP,3503208,ARARAQUARA,3500000,...,APOIO A PROJETOS DE INFRAESTRUTURA TURISTICA,0,APOIO A PROJETOS DE INFRAESTRUTURA TURISTICA -...,243750.00,0.0,0.0,0.0,243750.0,0.00,Sem informação
4,Sem informação,2014,Emenda Individual - Transferências com Finalid...,NaN,None,S/I,CAMPO LARGO - PR,4104204,CAMPO LARGO,4100000,...,ESTRUTURACAO DA REDE DE SERVICOS DE ATENCAO PR...,0,ESTRUTURACAO DA REDE DE SERVICOS DE ATENCAO BA...,1658333.94,0.0,0.0,0.0,527200.0,1131133.94,Sem informação
5,Sem informação,2014,Emenda Individual - Transferências com Finalid...,NaN,None,S/I,CAMPINAS DO SUL - RS,4303806,CAMPINAS DO SUL,4300000,...,APOIO A PROJETOS DE INFRAESTRUTURA TURISTICA,0,APOIO A PROJETOS DE INFRAESTRUTURA TURISTICA -...,243750.00,0.0,0.0,0.0,0.0,243750.00,Sem informação
6,Sem informação,2014,Emenda Individual - Transferências com Finalid...,NaN,None,S/I,MORENO - PE,2609402,MORENO,2600000,...,APOIO A INFRAESTRUTURA PARA A EDUCACAO BASICA,0,INFRAESTRUTURA PARA A EDUCACAO BASICA - DESPES...,700000.00,0.0,0.0,0.0,700000.0,0.00,Sem informação
7,Sem informação,2014,Emenda Individual - Transferências com Finalid...,NaN,None,S/I,NOVA FLORESTA - PB,2510105,NOVA FLORESTA,2500000,...,ESTRUTURACAO DA REDE DE SERVICOS DE ATENCAO PR...,0,ESTRUTURACAO DA REDE DE SERVICOS DE ATENCAO BA...,278700.00,0.0,0.0,0.0,0.0,278700.00,Sem informação
8,Sem informação,2014,Emenda Individual - Transferências com Finalid...,NaN,None,S/I,SANTIAGO DO SUL - SC,4215695,SANTIAGO DO SUL,4200000,...,FOMENTO AO SETOR AGROPECUARIO,0,FOMENTO AO SETOR AGROPECUARIO - DESPESAS DIVERSAS,300000.00,300000.0,300000.0,0.0,0.0,0.00,Sem informação
9,Sem informação,2014,Emenda Individual - Transferências com Finalid...,NaN,None,S/I,CIANORTE - PR,4105508,CIANORTE,4100000,...,APOIO A POLITICA NACIONAL DE DESENVOLVIMENTO U...,0,APOIO A POLITICA NACIONAL DE DESENVOLVIMENTO U...,245850.00,0.0,0.0,0.0,0.0,245850.00,Sem informação


Explique em uma frase o que você observa nas primeiras 10 linhas.

AS PRIMEIRAS LINHAS MOSTRAM O INICIO DOS DADOS DA TABELA, MOSTRANDO INDENTIFICADORES DE EMENDA, VALORES, DATAS, LOCALIZAÇÕES.

## 2) Quantas linhas e colunas o dataset possui?

In [7]:
df.shape

(88659, 29)

Explique o que esses números significam.

A QUANTIDADE DE LINHAS (88659), REPRESENTAM A QUANTIDADE DE EMENDAS, JA A QUANTIDADE DE COLUNAS (29), REPRESENTAM QUANTOS ATRIBUTOS FORAM DADOS AS EMENDAS

## 3) Liste os nomes das colunas e descreva brevemente o que você entende que cada coluna representa.

In [9]:
colunas= df.columns.tolist
print(colunas)

<bound method IndexOpsMixin.tolist of Index(['Código da Emenda', 'Ano da Emenda', 'Tipo de Emenda',
       'Código do Autor da Emenda', 'Nome do Autor da Emenda',
       'Número da emenda', 'Localidade de aplicação do recurso',
       'Código Município IBGE', 'Município', 'Código UF IBGE', 'UF', 'Região',
       'Código Função', 'Nome Função', 'Código Subfunção', 'Nome Subfunção',
       'Código Programa', 'Nome Programa', 'Código Ação', 'Nome Ação',
       'Código Plano Orçamentário', 'Nome Plano Orçamentário',
       'Valor Empenhado', 'Valor Liquidado', 'Valor Pago',
       'Valor Restos A Pagar Inscritos', 'Valor Restos A Pagar Cancelados',
       'Valor Restos A Pagar Pagos', 'Código da Emenda_1'],
      dtype='object')>


Escreva sua breve descrição das colunas, como se estivesse criando um dicionário de variáveis para que leitor entenda.

CÓDIGO DA EMENDA- IDENTIFICADOR UNICO PARA A EMENDA
TIPO DA EMENDA- EXIBE QUAL O TIPO DA EMENDA
ANO DA EMENDA- EXIBE DE QUAL ANO É A EMENDA
CODIGO MUNICIPIO IBGE- IDENTIFICADOR PARA O LOCAL DA EMENDA
CODIGO DO AUTOR DA EMENDA- IDENTIFICADOR PARA OS AUTORES DA EMENDA
NOME DO AUTOR DA EMENDA
NUMERO DA EMENDA- IDENTIFICADOR UNICO PARA A EMENDA
LOCALIDADE DE APLICAÇÃO DE RECURSO- LOCAL ONDE A EMENDA FOI APLICADA
MUNICIPIO- LOCAL DA EMENDA
NOME AÇÃO- NOME DA AÇÃO APLICADA PARA A EMENDA
CODIGO PLANO ORÇAMENTÁRIO- IDENTIFICADOR PARA O PLANO ORÇAMENTARIO
VALOR EMPENHADO- VALOR DEDICADO À EMENDA


## 4) Existem valores nulos? quantos?


In [14]:
df.isnull().sum().sum()

np.int64(31922)

## 5) Mostre os tipos das variáveis (se int, string, boolean, float...) e analise se alguma variável parece ter o tipo errado. Faça isso da maneira mais objetiva e otimizada possível.

In [15]:
df.dtypes

,0
Código da Emenda,object
Ano da Emenda,int64
Tipo de Emenda,object
Código do Autor da Emenda,float64
Nome do Autor da Emenda,object
Número da emenda,object
Localidade de aplicação do recurso,object
Código Município IBGE,object
Município,object
Código UF IBGE,int64


Se alguma coluna estiver com o tipo errado, escreva abaixo qual é e por que isso pode atrapalhar análises.

**Dica:** Se alguma variável tiver no tipo o nome "object", isso se refere a texto.

COLUNAS ERRADAS
Código do Autor da Emenda
Código Município IBGE
Código UF IBGE
Código Função/Código Subfunção, Código Programa, Código Ação, Código Plano Orçamentário

Atrapalha análises porque gera cálculos indevidos, a ordem pode ficar incorreta, e alguns agrupamentos sao comprometidos

## 6) Usando a coluna `Ano da Emenda` como base, descubra até que ano o dataset está atualizado (mostre as 8 primeiras linhas do df ordenado desc).

In [17]:
df.sort_values(by='Ano da Emenda',ascending=False).head(8)

,Código da Emenda,Ano da Emenda,Tipo de Emenda,Código do Autor da Emenda,Nome do Autor da Emenda,Número da emenda,Localidade de aplicação do recurso,Código Município IBGE,Município,Código UF IBGE,...,Nome Ação,Código Plano Orçamentário,Nome Plano Orçamentário,Valor Empenhado,Valor Liquidado,Valor Pago,Valor Restos A Pagar Inscritos,Valor Restos A Pagar Cancelados,Valor Restos A Pagar Pagos,Código da Emenda_1
88564,202541820003,2025,Emenda Individual - Transferências Especiais,4182.0,ZEQUINHA MARINHO,3,MÚLTIPLO,Sem informação,Múltiplo,1500000,...,TRANSFERENCIAS ESPECIAIS,0,TRANSFERENCIAS ESPECIAIS,12161000.00,12039390.00,12039390.00,0.0,0.0,0.0,202541820003
88595,202544840009,2025,Emenda Individual - Transferências com Finalid...,4484.0,ZUCCO,9,MÚLTIPLO,Sem informação,Múltiplo,4300000,...,DESENVOLVIMENTO DE ATIVIDADES E APOIO A PROGRA...,0,DESENVOLVIMENTO DE ATIVIDADES E APOIO A PROGRA...,1800000.00,36000.00,0.00,0.0,0.0,0.0,202544840009
88565,202541820004,2025,Emenda Individual - Transferências com Finalid...,4182.0,ZEQUINHA MARINHO,4,MÚLTIPLO,Sem informação,Múltiplo,1500000,...,FOMENTO AO SETOR AGROPECUARIO,0,FOMENTO AO SETOR AGROPECUARIO - DESPESAS DIVERSAS,4080113.09,1639405.41,1636170.92,0.0,0.0,0.0,202541820004
88566,202541820005,2025,Emenda Individual - Transferências com Finalid...,4182.0,ZEQUINHA MARINHO,5,PARÁ (UF),Sem informação,Sem informação,1500000,...,FOMENTO AO SETOR AGROPECUARIO,0,FOMENTO AO SETOR AGROPECUARIO - DESPESAS DIVERSAS,3000000.00,2978232.91,2978232.91,0.0,0.0,0.0,202541820005
88567,202541820006,2025,Emenda Individual - Transferências com Finalid...,4182.0,ZEQUINHA MARINHO,6,PARÁ (UF),Sem informação,Sem informação,1500000,...,APOIO AS INICIATIVAS DE PROMOCAO DA IGUALDADE ...,0,IGUALDADE DE DIREITOS E AUTONOMIA ECONOMICA DA...,1000000.00,0.00,0.00,0.0,0.0,0.0,202541820006
88584,202544940007,2025,Emenda Individual - Transferências com Finalid...,4494.0,ZEZINHO BARBARY,7,ACRE (UF),Sem informação,Sem informação,1200000,...,AQUISICAO E DISTRIBUICAO DE ALIMENTOS DA AGRIC...,0,AQUISICAO E DISTRIBUICAO DE ALIMENTOS DA AGRIC...,600000.00,0.00,0.00,0.0,0.0,0.0,202544940007
88585,202544940008,2025,Emenda Individual - Transferências com Finalid...,4494.0,ZEZINHO BARBARY,8,Nacional,Sem informação,Sem informação,-1,...,ADEQUACAO DE ATIVOS DE INFRAESTRUTURA DAS ORGA...,0,ADEQUACAO DE ATIVOS DE INFRAESTRUTURA DAS ORGA...,499871.46,183600.86,168502.47,0.0,0.0,0.0,202544940008
88586,202544940009,2025,Emenda Individual - Transferências com Finalid...,4494.0,ZEZINHO BARBARY,9,Nacional,Sem informação,Sem informação,-1,...,ASSISTENCIA MEDICA QUALIFICADA E GRATUITA A TO...,0,ASSISTENCIA MEDICA QUALIFICADA E GRATUITA A TO...,200000.00,200000.00,200000.00,0.0,0.0,0.0,202544940009


Explique: qual é o ano mais recente presente no dataset?

2025

## 7) execute um  `describe()` e explique uma conclusão simples que você consegue tirar a partir das estatísticas (se algumas colunas deveriam aparecer no describe mas não aparecem, indique o motivo que você acredita que esteja causando esse problema).

In [18]:
df.describe()

,Ano da Emenda,Código do Autor da Emenda,Código UF IBGE,Valor Empenhado,Valor Liquidado,Valor Pago,Valor Restos A Pagar Inscritos,Valor Restos A Pagar Cancelados,Valor Restos A Pagar Pagos
count,88659.000000,72698.000000,8.865900e+04,8.865900e+04,8.865900e+04,8.865900e+04,8.865900e+04,8.865900e+04,8.865900e+04
mean,2019.154288,3345.556439,2.943364e+06,3.359837e+06,1.893686e+06,1.871780e+06,1.232317e+06,1.477261e+05,9.629035e+05
std,3.408250,1394.534620,1.264778e+06,4.233214e+07,3.304432e+07,3.300694e+07,3.544986e+07,3.008922e+06,1.377985e+07
min,2014.000000,-99.000000,-1.400000e+01,0.000000e+00,0.000000e+00,0.000000e+00,-6.351103e+05,0.000000e+00,-1.200000e+05
25%,2016.000000,2707.000000,2.300000e+06,2.500000e+05,0.000000e+00,0.000000e+00,0.000000e+00,0.000000e+00,0.000000e+00
50%,2019.000000,3298.000000,3.100000e+06,5.411500e+05,0.000000e+00,0.000000e+00,0.000000e+00,0.000000e+00,1.500000e+05
75%,2022.000000,3994.000000,3.500000e+06,1.799922e+06,5.000000e+05,5.000000e+05,2.000000e+05,0.000000e+00,5.000000e+05
max,2025.000000,9235.000000,5.300000e+06,5.019577e+09,5.019577e+09,5.019577e+09,6.008343e+09,4.276822e+08,2.114019e+09


Algumas colunas nao aparecem porque estao como object


## 8) Mostre as últimas 5 linhas do dataset e escreva se a última linha parece completa (sem muitos nulos) ou não.

In [19]:
df.tail(5)

,Código da Emenda,Ano da Emenda,Tipo de Emenda,Código do Autor da Emenda,Nome do Autor da Emenda,Número da emenda,Localidade de aplicação do recurso,Código Município IBGE,Município,Código UF IBGE,...,Nome Ação,Código Plano Orçamentário,Nome Plano Orçamentário,Valor Empenhado,Valor Liquidado,Valor Pago,Valor Restos A Pagar Inscritos,Valor Restos A Pagar Cancelados,Valor Restos A Pagar Pagos,Código da Emenda_1
88654,201811420017,2018,Emenda Individual - Transferências com Finalid...,1142.0,ZE GERALDO,17,PARÁ (UF),Sem informação,Sem informação,1500000,...,CONSTRUCAO E AMPLIACAO DE UNIDADES OPERACIONAI...,EIND,EMENDA INDIVIDUAL,500000.0,0.0,0.0,0.0,369387.36,130612.64,201811420017
88655,201827690001,2018,Emenda Individual - Transferências com Finalid...,2769.0,ZE SILVA,1,MÚLTIPLO,Sem informação,Múltiplo,3100000,...,ESTRUTURACAO DA REDE DE SERVICOS DE ATENCAO PR...,EIND,EMENDA INDIVIDUAL,499975.0,499975.0,499975.0,0.0,0.00,0.00,201827690001
88656,201827690002,2018,Emenda Individual - Transferências com Finalid...,2769.0,ZE SILVA,2,ITURAMA - MG,3134400,ITURAMA,3100000,...,APOIO A PROJETOS DE DESENVOLVIMENTO SUSTENTAVE...,EIND,EMENDA INDIVIDUAL,200000.0,0.0,0.0,11058.0,0.00,188942.00,201827690002
88657,201827690003,2018,Emenda Individual - Transferências com Finalid...,2769.0,ZE SILVA,3,ITURAMA - MG,3134400,ITURAMA,3100000,...,ESTRUTURACAO DE UNIDADES DE ATENCAO ESPECIALIZ...,EIND,EMENDA INDIVIDUAL,199970.0,199970.0,199970.0,0.0,0.00,0.00,201827690003
88658,201827690004,2018,Emenda Individual - Transferências com Finalid...,2769.0,ZE SILVA,4,POUSO ALEGRE - MG,3152501,POUSO ALEGRE,3100000,...,APOIO A REALIZACAO DE EVENTOS DE ESPORTE AMADO...,EIND,EMENDA INDIVIDUAL,100000.0,0.0,0.0,0.0,6617.47,93382.53,201827690004


Interprete a completude da última linha.

A ultima linha esta completa
Todos os principais campos estão preenchidos
Os valores Financeiros sao numericos
O codigo da emenda esta presente
A estrutura esta identica as demais linhas



## 9) existem valores duplicados no dataframe? Caso positivo, quantos?

In [20]:
df.duplicated().sum()

np.int64(63)

63 linhas dupkicadas